core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -761,92 +761,107 @@ sealed trait ZSTM[-R, +E, +A] extends Serializable { self =>
     type Erased = ZSTM[Any, Any, Any]
     type Cont   = Any => Erased
 
-    val contStack     = Stack[Cont]()
-    val envStack      = Stack[ZEnvironment[Any]](r0)
-    var onCommitStack = Stack[ZIO[Any, Nothing, Any]]()
+    var contStack     = List.empty[Cont]
+    var onCommitStack = List.empty[ZIO[Any, Nothing, Any]]
+    var env           = r0.asInstanceOf[ZEnvironment[Any]]
     var exit          = null.asInstanceOf[TExit[Any, Any]]
     var curr          = self.asInstanceOf[Erased]
     var opCount       = 0
 
-    def unwindStack(error: Any, isRetry: Boolean): Erased = {
+    def unwindStack(stack: List[Cont], error: Any, isRetry: Boolean): (List[Cont], Erased) = {
       var result = null.asInstanceOf[Erased]
+      var rem    = stack
 
-      while (!contStack.isEmpty && (result eq null)) {
-        contStack.pop() match {
+      while ((rem ne Nil) && (result eq null)) {
+        rem.head match {
           case OnFailure(_, onFailure) => if (!isRetry) result = onFailure.asInstanceOf[Cont].apply(error)
           case OnRetry(_, onRetry)     => if (isRetry) result = onRetry
           case _                       =>
         }
+        rem = rem.tail
       }
 
-      result
+      (rem, result)
     }
 
     while (exit eq null) {
       if (opCount == YieldOpCount) {
-        if (isInvalid(journal)) exit = TExit.Retry
+        if (journal.isInvalid) exit = TExit.Retry
         else opCount = 0
       } else {
         try {
-          (curr.tag: @annotation.switch) match {
-            case Tags.Effect =>
-              val effect = curr.asInstanceOf[Effect[Any, Any, Any]]
-              val a      = effect.f(journal, fiberId, envStack.peek())
-              if (contStack.isEmpty) exit = TExit.Succeed(a, onCommitStack.toList) else curr = contStack.pop()(a)
-
-            case Tags.OnSuccess =>
-              val onSuccess = curr.asInstanceOf[OnSuccess[Any, Any, Any, Any]]
-              contStack.push(onSuccess.k)
+          curr match {
+            case effect: Effect[Any, Any, Any] =>
+              val a = effect.f(journal, fiberId, env)
+              if (contStack eq Nil) exit = TExit.Succeed(a, onCommitStack)
+              else {
+                curr = contStack.head(a)
+                contStack = contStack.tail
+              }
+
+            case onSuccess: OnSuccess[Any, Any, Any, Any] =>
+              contStack ::= onSuccess.k
```

Comment:
what is that `::=`? 🤔

Reply:
Syntactic sugar for `contStack = onSuccess.k :: contStack` (contStack is List `var`)